README.md

```diff
@@ -1,16 +1,204 @@
-# UCAN Invocation
+# UCAN Invocation Specification v0.1.0
 
 ## Editors
 
-* _
+* [Brooklyn Zelenka](https://github.com/expede/), [Fission](https://fission.codes/)
 
 ## Authors
 
-* _
+* [Brooklyn Zelenka](https://github.com/expede/), [Fission](https://fission.codes/)
+* [Irakli Gozalishvili](https://github.com/Gozala), [DAG House](https://dag.house/)
 
-# 0. Abstract
+## Depends On
+
+* [`ucan-ipld`](https://github.com/ucan-wg/ucan-ipld/)
+
+# 0 Abstract
 
 ## Language
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
 
+# 1 Introduction
+
+## Separation
+
+Why separate the UCAN from the invocation format? Surely the UCAN itself already contains all the infomation required.
+
+I would argue that this is not the case for a few reasons.
+
+1. Authority is not intent to execute. Authority transfer is first-class in UCAN, other actions are not
+2. Mixing the two levels means that invocation info will live with the token (as a proof) on subdelegations
+3. Other detail, such as scheduling
+
+A job request may be picked up by anyone, PRIOR to UCAN delegation. A handshake or matchmaking may need to be performed.
+
+## Authority Is Not Intention
+
+Or to put it another way: "just because you can, doens't mean you should". Granting a UCAN to a peer means that they are allowed to perform some actions for a period of time. This is a feature not a bug, but also says nothing about the intention of when it should be run. I may grant a collaborative process the ability to perform actions on an ongoing basis (hmm, but vioating POLA)
+
+I could preload the service with `n` UCANs with narrow time windows and `nbf`s in the future. That would certainly be very secure, but it would be less convenient since I need to come online to issue more or to 
+
+# 2 Envelope
+
+UCAN uses a capabilities model. The 
+
+* UCANTO
+* zcap
+* IPVM
+
+
+
+
+Specify Roles
+
+
+
+
+## 2.1 IPLD
+
+``` ipldsch
+type Invocation struct {
+  inv [&UCAN] -- To invoke
+  v   SemVer  -- Version
+  nnc String  -- Nonce
+  sig Bytes   -- Signature
+}
+
+type SemVer struct {
+  ma Integer
+  mi Integer
+  pa Integer
+  ta optional nullable String
+}
+```
+
+## 2.2 Exmaple
+
+ ``` json
+{
+  "ucan/invoke": [ "bafyLeft", "bafyRight", "bafyEnd" ]
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "sig": 0xCOFFEE
+}
+```
+
+# 3 Receipt & Attestation
+
+An invocation receipt is a claim about what the output of an invocation is. A receipt MUST be attested via signature of the principal (the audience of the associated invocation).
+
+Note that this does not guarantee correctness of the result! The level of this statement's veracity MUST be ony taken that the signer claims this to be a fact.
+
+The receipt MUST be signed with by the `aud` from the UCAN.
```

Comment:
I find this to requirement to be difficult to meet. In our setup actor doing the work would not have access to the key to sign the result, but it would have some delegation from the `aud` to act on it's behalf.

I think it would be a good idea to consider such a setup and how delegates could sign on behalf of `aud`.

Reply:
The idea is that if a subdelegate is going to do the work, you need to create an invocation for them

```
alice --[delegate]--> bob --[delegate]--> carol --[delegate]--> dan
                                &                     &
                              Invoke                Invoke
```

Here we have both an delegation chain (the right to do the thing), and a wrapper invocation. Bob asks Carol to do the work. She doens't have access to everything required for the job, so she subdelegates a partial job to Dan and wraps it in another invocation.

Does that not match your picture?

Reply:
Problem is that invocation will target web3.storage which has some did:key. But service that receives that invocation will not have access to that key, so it will not be able to sign receipt with it.

Service could have some delegation from web3.storage key as an authorization to act on it’s behalf, which it will get earlier than invocation.

This is more or less what I’ve been trying to accomplish with pipelining/forwarding threads. I

Reply:
Ooooooh interesting @Gozala! You're trying to do that statelessly. The other way of modeling that is of course with a (stateful) DID document.

The problem with not aligning the audience in the invocation is that if you need to coordinate with an external service, you won't be able to prove the chain. For example, let's say that Fission starts using web3.storage for storage, and web3.storage uses Fission for DNSLink management. We would have no way to prove that the user intends to update their DNSLink if they started a request from your service (to abstract away the fact that there's many services here).

I agree that the forwarding concept would alleviate this! Also agree that "sure we'll run anything aimed at that DID" also works for a subset of cases where you're handing the entire request. If you want a service to openly interoperate, "the chain must be unbroken".

So: if we ship the forwarding/powerbox feature, does that make invocation better for you?

Reply:
> Ooooooh interesting @Gozala! You're trying to do that statelessly. The other way of modeling that is of course with a (stateful) DID document.

You could do with did docs but that indeed introduces state and has additional tradeoffs that I’d like to avoid. 

I’m not trying to make it stateless, but rather trying to capture state in the receipt so it could be verified even after state change

> The problem with not aligning the audience in the invocation is that if you need to coordinate with an external service, you won't be able to prove the chain. For example, let's say that Fission starts using web3.storage for storage, and web3.storage uses Fission for DNSLink management. We would have no way to prove that the user intends to update their DNSLink if they started a request from your service (to abstract away the fact that there's many services here).

I think we misunderstand each other. All I’m proposing is just like agent could delegate invocation to another agent, it should be possible to delegate execution to another agent. 

In other words I could delegate to you file upload capability. But so should upload service be able to delegate providing upload capability to another service. That way agent invoking capability does not need to know which upload service web3.storage relies on.

It is possible to accomplish above with active forwarding of delegations  where our service accepts invocation and then forwards (redelegates) it to desired handler. But that implies signing things at every request and access to the keys.

Ideally it would be possible to just route request to a desired handler, who could sign receipt with own key and provide a proof chain warranting it to handle such requests on behalf of the `aud` service.


> So: if we ship the forwarding/powerbox feature, does that make invocation better for you?

That might be a way to address it, but there also might be different ways to address this specific problem without changing UCANs themselves

Reply:
> I’m not trying to make it stateless, but rather trying to capture state in the receipt so it could be verified even after state change

Sorry, this is what I mean by "stateless" — in the same sense REST. It doesn't make reference to ambient external state, but everything required to fulfill the request is passed along together.

> That way agent invoking capability does not need to know which upload service web3.storage relies on.

Absolutely! My understanding about this scenario is that web3.storage would be unable to _further delegate to external services_. Is that correct?

```
Want:
Alice -> Bob -> web3.storage[bot1, bot2, bot3] -> fission.codes

But have these disconnected:

Alice -> Bob -> web3.storage
bot2 -> fission.codes
```

This scenario would be impossible, because web3.storage would not be able to subdelegate to fission.codes. 

This can be fixed with your forwarding/powerbox idea to connect the two disconnected chains:

```
Alice -> Bob -> web3.storage

web3.storage =[all]=> bot1
web3.storage =[all]=> bot2
web3.storage =[all]=> bot3

bot2 -> fission.codes

Which pictorially composes something like:

Alice -> Bob -> web3.storage
                web3.storage ==[all]==> bot2
                                        bot2 -> fission.codes
```

This is my current picture/plan for 0.10 ☝️ Does that match yours?